mava/networks/sable_network.py

```diff
@@ -0,0 +1,547 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import partial
+from typing import Any, Optional, Tuple
+
+import chex
+import jax
+import jax.numpy as jnp
+from flax import linen as nn
+from flax.linen.initializers import orthogonal
+from omegaconf import DictConfig
+
+from mava.networks.retention import MultiScaleRetention
+from mava.networks.torsos import SwiGLU
+from mava.networks.utils.sable.discrete_trainer_executor import *  # noqa
+from mava.systems.sable.types import HiddenStates, SableNetworkConfig
+from mava.types import Observation
+
+
+class EncodeBlock(nn.Module):
+    """Sable encoder block."""
+
+    net_config: SableNetworkConfig
+    memory_config: DictConfig
+    n_agents: int
+
+    def setup(self) -> None:
+        # Initialize the RMSNorm layer normalization
+        self.ln1 = nn.RMSNorm()
+        self.ln2 = nn.RMSNorm()
+
+        # Initialize the MultiScaleRetention
+        self.retn = MultiScaleRetention(
+            embed_dim=self.net_config.embed_dim,
+            n_head=self.net_config.n_head,
+            n_agents=self.n_agents,
+            full_self_retention=True,  # Full retention for the encoder
+            memory_config=self.memory_config,
+            decay_scaling_factor=self.memory_config.decay_scaling_factor,
+        )
+
+        # Initialize SwiGLU feedforward network
+        self.ffn = SwiGLU(self.net_config.embed_dim, self.net_config.embed_dim)
```

Comment:
Maybe I'm being too harsh here, but I feel these kinds of comments are unnecessary given the code is self explainatory

```suggestion
        self.ln1 = nn.RMSNorm()
        self.ln2 = nn.RMSNorm()

        self.retn = MultiScaleRetention(
            embed_dim=self.net_config.embed_dim,
            n_head=self.net_config.n_head,
            n_agents=self.n_agents,
            full_self_retention=True,  # Full retention for the encoder
            memory_config=self.memory_config,
            decay_scaling_factor=self.memory_config.decay_scaling_factor,
        )

        self.ffn = SwiGLU(self.net_config.embed_dim, self.net_config.embed_dim)
```